File: Python/textbook_track/chapter08/ba8b.py
# Compute the Squared Error Distortion
import functools as ft
import math
import sys


def convert_to_numbers(line, converter_func):
    return [converter_func(elem) for elem in line.split()]


def read_points(reader):
    return [convert_to_numbers(line, float) for line in reader]


def euclidean_distance(p, q):
    return math.sqrt(sum((p_i - q_i)**2 for p_i, q_i in zip(p, q)))


def distance_from_centers(point, centers):
    distance_from_point = ft.partial(euclidean_distance, q=point)
    return min(map(distance_from_point, centers))


def calc_distortion(points, centers):
    n = len(points)
    return sum(distance_from_centers(point, centers) ** 2 for point in points) / n


def main():
    reader = sys.stdin
    k, _ = convert_to_numbers(next(reader), int)
    centers = [convert_to_numbers(next(reader), float) for _ in range(k)]
    _ = next(reader)
    points = read_points(reader)
    result = calc_distortion(points, centers)
    print(result)


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter08/BA8B.scala
package TextbookTrack.Chapter08

object BA8B {
  import scala.reflect.ClassTag

  type Point = List[Double]

  private def convertToNumber[T: ClassTag](line: String, converter: String => T): List[T] =
    line.split(" ").map(converter).toList

  private def readParameters(line: String): (Int, Int) = convertToNumber(line, _.toInt) match {
    case List(k, m) => (k, m)
    case _ => throw new Exception("Unexpected number of parameters.")
  }

  private def euclideanDistance(p: Point, q: Point): Double =
    math.sqrt(p.lazyZip(q).foldLeft(0.0){ case (d, (pi, qi)) => d + math.pow(pi - qi, 2) })

  private def distanceFromCenters(point: Point, centers: List[Point]): Double =
    centers.map(euclideanDistance(_, point)).min

  def calcDistortion(points: List[Point], centers: List[Point]): Double =
    points.foldLeft(0.0)((d, point) => d + math.pow(distanceFromCenters(point, centers), 2)) / points.length

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val (k, _): (Int, Int) = readParameters(reader.next())
    val centers: List[Point] = reader.take(k).map(convertToNumber(_, _.toDouble)).toList
    val _: String = reader.next()
    val points: List[Point] = reader.map(convertToNumber(_, _.toDouble)).toList
    val result: Double = calcDistortion(points, centers)
    println(result)
  }
}
